lib/iris/_merge.py

```diff
@@ -1613,22 +1644,32 @@ def _build_coordinates(self):
             # Attempt to build a DimCoord and add it to the cube. If this
             # fails e.g it's non-monontic or multi-dimensional or non-numeric,
             # then build an AuxCoord.
-            try:
-                coord = iris.coords.DimCoord(
-                    template.points, bounds=template.bounds, **template.kwargs
-                )
-                if len(template.dims) == 1 and template.dims[0] not in covered_dims:
-                    dim_coords_and_dims.append(_CoordAndDims(coord, template.dims))
-                    covered_dims.append(template.dims[0])
-                else:
-                    aux_coords_and_dims.append(_CoordAndDims(coord, template.dims))
-            except ValueError:
+
+            # Check here whether points are masked? If so then it has to be an AuxCoord
+            if np.ma.is_masked(template.points):
                 # kwarg not applicable to AuxCoord.
                 template.kwargs.pop("circular", None)
                 coord = iris.coords.AuxCoord(
                     template.points, bounds=template.bounds, **template.kwargs
                 )
                 aux_coords_and_dims.append(_CoordAndDims(coord, template.dims))
+            else:
+                try:
+                    coord = iris.coords.DimCoord(
+                        template.points, bounds=template.bounds, **template.kwargs
+                    )
+                    if len(template.dims) == 1 and template.dims[0] not in covered_dims:
+                        dim_coords_and_dims.append(_CoordAndDims(coord, template.dims))
+                        covered_dims.append(template.dims[0])
+                    else:
+                        aux_coords_and_dims.append(_CoordAndDims(coord, template.dims))
+                except (ValueError, TypeError):
+                    # kwarg not applicable to AuxCoord.
+                    template.kwargs.pop("circular", None)
+                    coord = iris.coords.AuxCoord(
+                        template.points, bounds=template.bounds, **template.kwargs
+                    )
+                    aux_coords_and_dims.append(_CoordAndDims(coord, template.dims))
```

Comment:
I would prefer to see this in a little inner function, rather than copied twice.